controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,414 @@
+			Expect(!result.CancelRequest && err == nil).To(BeTrue())
+		}
+		release.SetDeletionTimestamp(&metav1.Time{Time: time.Now()})
+
+		// delete the pipelineRun before the release so it should thrown an
+		// error on finalizing the release
+		pipelineRun, _ := cacheAdapter.getReleasePipelineRun()
+		Expect(cacheClient.Delete(ctx, pipelineRun)).Should(Succeed())
+		Expect(cacheClient.Delete(ctx, release)).Should(Succeed())
+
+		result, err := cacheAdapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err != nil).To(BeTrue())
+	})
+
+	It("can create a ReleasePipelineRun within the given adapter", func() {
+		applicationSnapshot.TypeMeta.Kind = "applicationSnapshot"
+		Expect(noCacheClient.Update(ctx, applicationSnapshot)).Should(Succeed())
+		pipelineRun, err := adapter.createReleasePipelineRun(
+			releaseStrategy,
+			applicationSnapshot)
+		Expect(pipelineRun != nil && err == nil).To(BeTrue())
+		Expect(noCacheClient.Delete(ctx, pipelineRun)).Should(Succeed())
+	})
+
+	It("can finalize (delete) the releasePipelineRun just processed", func() {
+		result, err := cacheAdapter.EnsureFinalizersAreCalled()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureFinalizerIsAdded()
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}).Should(BeTrue())
+		result, err = cacheAdapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		err = cacheAdapter.finalizeRelease()
+		Expect(err).Should(Succeed())
+
+		pipelineRun, _ := cacheAdapter.getReleasePipelineRun()
+		Expect(cacheClient.Delete(ctx, pipelineRun)).ShouldNot(Succeed())
+	})
+
+	It("can return the ApplicationSnapshot from the release on the given adapter", func() {
+		snapshot, err := adapter.getApplicationSnapshot()
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(snapshot)).
+			To(Equal(reflect.TypeOf(&appstudioshared.ApplicationSnapshot{})))
+
+		// should error when applicationSnapshot is missing
+		Expect(noCacheClient.Delete(ctx, applicationSnapshot)).Should(Succeed())
+		_, err = adapter.getApplicationSnapshot()
+		Expect(err).Should(HaveOccurred())
+	})
+
+	It("can return the releaseLink from the release within the given adapter", func() {
+		link, err := adapter.getReleaseLink()
+		Expect(err).Should(Succeed())
+		Expect(reflect.TypeOf(link)).
+			To(Equal(reflect.TypeOf(&appstudiov1alpha1.ReleaseLink{})))
+
+		// shold err when releaseLink does not exist
+		noCacheClient.Delete(ctx, releaseLink)
+		_, err = adapter.getReleaseLink()
+		Expect(err).Should(HaveOccurred())
+	})
+
+	It("can return the pipelineRun from the release within the give adapter", func() {
+		Eventually(func() bool {
+			result, err := cacheAdapter.EnsureReleasePipelineRunExists()
```

Comment:
Not sure you even need this. You create a Release in the `BeforeEach` and that is going to create a PipelineRun already. I'd say that you can add the `getReleasePipelineRun` call here and wait for it to succeed.

Reply:
Got it. I thought beforehand that I would need to explicitly call the `Reconcile()` but indeed it is called when the object is handled by the cluster. I have lots of this. I will remove.